src/silo/query_engine/actions/aggregated.cpp

```diff
@@ -0,0 +1,304 @@
+Aggregated::Aggregated(
+   std::vector<std::string> group_by_fields,
+   std::vector<std::string> order_by_fields,
+   std::optional<uint32_t> limit
+)
+    : group_by_fields(std::move(group_by_fields)),
+      order_by_fields(std::move(order_by_fields)),
+      limit(limit) {}
+
+QueryResult Aggregated::execute(
+   const Database& database,
+   std::vector<OperatorResult> bitmap_filters
+) const {
+   if (group_by_fields.empty()) {
+      uint32_t count = 0;
+      for (const auto& filter : bitmap_filters) {
+         count += filter->cardinality();
+      };
+      std::map<std::string, std::variant<std::string, uint32_t>> tuple_fields;
+      tuple_fields["count"] = count;
+      return QueryResult{std::vector<AggregationResult>{{tuple_fields}}};
+   }
+   if (group_by_fields.size() == 1) {
+      const std::string group_by_field = group_by_fields[0];
+      const auto& metadata = database.database_config.getMetadata(group_by_field);
+      if (metadata.name == database.database_config.schema.primary_key) {
+         throw QueryParseException("Cannot group by primary key field: '" + group_by_field + "'");
+      }
+      // TODO(taepper) optimize when equal to partition_by field
+      // TODO(taepper) optimize single field groupby
+   }
+   std::vector<config::DatabaseMetadata> group_by_metadata;
+   for (const std::string& group_by_field : group_by_fields) {
+      const auto& metadata = database.database_config.getMetadata(group_by_field);
+      group_by_metadata.push_back(metadata);
+   }
+
+   std::vector<TupleAccessor> order_by_definition;
+   bool randomize_order = false;
+   for (const std::string& order_by_field : order_by_fields) {
+      if (order_by_field == "random()") {
+         randomize_order = true;
+         continue;
+      }
+      order_by_definition.push_back(parseOrderByField(order_by_field, group_by_metadata));
+   }
+
+   tbb::enumerable_thread_specific<std::unordered_map<Tuple, uint32_t>> maps;
+
+   tbb::parallel_for(
+      tbb::blocked_range<uint32_t>(0, database.partitions.size()),
+      [&](tbb::blocked_range<uint32_t> range) {
+         std::unordered_map<Tuple, uint32_t>& map = maps.local();
+         for (uint32_t partition_id = range.begin(); partition_id != range.end(); ++partition_id) {
+            const DatabasePartition& partition = database.partitions[partition_id];
+            for (const uint32_t sequence_id : *bitmap_filters[partition_id]) {
+               ++map[Tuple(sequence_id, partition.meta_store, group_by_metadata)];
+            }
+         }
+      }
+   );
+   std::unordered_map<Tuple, uint32_t> final_map;
+   for (auto& map : maps) {
+      for (auto& [key, value] : map) {
+         final_map[key] += value;
+      }
+   }
+   std::vector<Tuple> tuples;
+   for (auto& [tuple, count] : final_map) {
+      tuples.push_back(tuple);
+      tuples.back().addField(count);
+   }
+   group_by_metadata.push_back({"count", config::ValueType::NUMBER, false});
```

Comment:
This looks hacky^^ Is it worth creating a separate struct, so that we don't have to fall back on `DatabaseMetadata` here for a column that doesn't exist?

Reply:
The metadata is the descriptor of what values are present in the tuple. This is by far the simplest solution

Reply:
The metadata is the descriptor of which values are present in metadata columns. Tuples are not metadata columns (they only originate from them).

IMO: Not confusing developers is a lot more important than the "simple solution". A new struct generates just a little more code and should not have performance impact.

Reply:
But what would be in that struct? Another variant of either a column value or synthetic types that are created during the aggregation?

Reply:
Yes this would work and was the initial solution, but what if the user orders by count? Which is a common use-case imo

Reply:
That is why I needed to seamlessly include the `count` variable into the tuple, otherwise *all* later logic would need to hard code the count column and it was not feasible to create a working solution

Reply:
But let's go for "Either hardcode it there, that we always retrieve `count`". This should also work with ordering by count, doesn't it?

Reply:
I do not think that is easily possible.

But what is the problem with the current solution? This is nothing out of the ordinary

Reply:
The current solution is very hacky. Obvious code makes it easier for future developers (including us in a couple of weeks).